### PR TITLE
IC-6: GetUser implementation

### PR DESCRIPTION
**Pull request by sobuse**
Get user controller

### CODE REVIEW

File: src/InstagramClone.Api/Controllers/UsersController.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly InstagramCloneDbContext _context;
+        public UsersController(InstagramCloneDbContext context)
+        {
+            _context = context;
+        }
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<User>>> GetUsers()
+        {
+            return await _context.Users.ToListAsync();
+        }
```
Comment (ehimah) [inline src/InstagramClone.Api/Controllers/UsersController.cs:25]: is this part of the specification?

Comment (sobuse) [reply src/InstagramClone.Api/Controllers/UsersController.cs]: No, I have taken it down

File: src/InstagramClone.Api/Controllers/UsersController.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+
+        //  api/users/:id
+        [HttpGet("{Id}")]
+        public async Task<ActionResult<User>> GetUser(Guid id)
+        {
+            return await _context.Users.FindAsync(id);
+
+        }
```
[the remark this answers is not in the document]
Comment (sobuse) [reply src/InstagramClone.Api/Controllers/UsersController.cs]: Ok, Done

**Review by ehimah — CHANGES_REQUESTED** [review]
I think the problem with your routing is that you haven't enabled attribute routing. 

See this article to see how to go about it. https://docs.microsoft.com/en-us/aspnet/web-api/overview/web-api-routing-and-actions/attribute-routing-in-web-api-2

You may need to introduce a `WebApiConfig` class.

**Comment by sobuse** [comment]
I have introduced the web config class and I still get access to my controller this is what I am getting.
25 06 2022_01 26 03_REC
r

File: webConfig.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
```
Comment (ehimah) [inline webConfig.cs:5]: add ```using System.Web.Http;```

File: src/InstagramClone.Api/Startup.cs [hunk 4]
```diff
@@ -43,7 +44,24 @@ private static void Configure(WebApplication app)
 
             app.UseRouting();
 
+            app.UseEndpoints(endpoints =>
+            {
+                endpoints.MapControllerRoute(
+                    name: "default",
+                    pattern: "{controller=Home}/{action=Index}/{id?}");
+            });
```
Comment (ehimah) [inline src/InstagramClone.Api/Startup.cs:52]: this call is duplicated below so delete this one and leave the one below.

File: src/InstagramClone.Api/Startup.cs [hunk 5]
```diff
@@ -25,7 +26,7 @@ private static void ConfigureService(WebApplicationBuilder builder)
                 config.ReturnHttpNotAcceptable = true;
             }).AddXmlDataContractSerializerFormatters();
 
-           // builder.Services.AddControllers();
+             // builder.Services.AddControllers();
```
Comment (ehimah) [inline src/InstagramClone.Api/Startup.cs:29]: why did you comment out the controllers configuration? how will asp.net see and register your controllers?

File: src/InstagramClone.Api/Startup.cs [hunk 6]
```diff
@@ -43,7 +44,24 @@ private static void Configure(WebApplication app)
...
+            });
+            app.UseEndpoints(endpoints =>
+            {
+                endpoints.MapControllers();
+            });
+
+            //app.UseRouting(route =>
+            //{
+            //    route.MapControllers();
+            //});
```
Comment (ehimah) [inline src/InstagramClone.Api/Startup.cs:61]: cleanup

File: src/InstagramClone.Api/Startup.cs (hunk 4 above)
Comment (sobuse) [reply src/InstagramClone.Api/Startup.cs:52]: ok.

File: src/InstagramClone.Api/Startup.cs (hunk 5 above)
Comment (sobuse) [reply src/InstagramClone.Api/Startup.cs:29]: ooh, I missed it up.

File: src/InstagramClone.Api/Startup.cs (hunk 6 above)
Comment (sobuse) [reply src/InstagramClone.Api/Startup.cs:61]: Ok.

File: src/InstagramClone.Api/Controllers/UsersController.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        private readonly InstagramCloneDbContext _context;
+        public UsersController(InstagramCloneDbContext context)
+        {
+            _context = context;
+        }
+        
+        [HttpGet("{Id}")]
+        public async Task<ActionResult<User>> GetUser(Guid id)
+        {
```
Comment (ehimah) [inline src/InstagramClone.Api/Controllers/UsersController.cs:24]: can you please research the appropriate return type(s) for web API controller methods and update your action method to use it here.

read https://docs.microsoft.com/en-us/aspnet/web-api/overview/getting-started-with-aspnet-web-api/action-results#ihttpactionresult

...and consider returning an `IHttpActionResult` result

File: src/InstagramClone.Api/Controllers/UsersController.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    [Route("api/[controller]")]
+    public class UsersController : ControllerBase
+    {
+        private readonly InstagramCloneDbContext _context;
+        public UsersController(InstagramCloneDbContext context)
+        {
+            _context = context;
+        }
+        
+        [HttpGet("{Id}")]
```
Comment (ehimah) [inline src/InstagramClone.Api/Controllers/UsersController.cs:22]: you need to check if this is case-sensitive. 

is `"{Id}"` the same as `"{id}"`?

File: src/InstagramClone.Api/Startup.cs [hunk 9]
```diff
@@ -13,26 +15,47 @@ public static WebApplication InitialIzeApp(string[] args)
...
         private static void ConfigureService(WebApplicationBuilder builder)
         {
             
+            builder.Services.AddControllers(config =>
+            {
+                config.RespectBrowserAcceptHeader = true;
+                config.ReturnHttpNotAcceptable = true;
+            }).AddXmlDataContractSerializerFormatters();
+
+             builder.Services.AddControllers();
```
Comment (ehimah) [inline src/InstagramClone.Api/Startup.cs:29]: something is wrong here and I have given you similar feedback about duplicates recently. 

fix up

File: src/InstagramClone.Api/Startup.cs [hunk 10]
```diff
@@ -13,26 +15,47 @@ public static WebApplication InitialIzeApp(string[] args)
...
+
 
         private static void ConfigureService(WebApplicationBuilder builder)
         {
             
+            builder.Services.AddControllers(config =>
+            {
+                config.RespectBrowserAcceptHeader = true;
+                config.ReturnHttpNotAcceptable = true;
+            }).AddXmlDataContractSerializerFormatters();
```
Comment (ehimah) [inline src/InstagramClone.Api/Startup.cs:27]: but really, what do these lines do? can you explain?

File: src/InstagramClone.Api/Controllers/UsersController.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ehimah) [reply src/InstagramClone.Api/Controllers/UsersController.cs]: What is the appropriate error code when a resource is not found on the server?

**Comment by ehimah** [comment]
@sobuse  what's the update on this? can we work to close this today?

File: src/InstagramClone.Api/Controllers/UsersController.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ehimah) [reply src/InstagramClone.Api/Controllers/UsersController.cs]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

list of HTTP status codes

File: src/InstagramClone.Api/Controllers/UsersController.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+        [Route("{id:Guid}")]
+        public IHttpActionResult GetUser(Guid id)
+        {
+            var user = _context.Users.Find(id);
+            if (user == null)
+            {
+                return NotFound();
+            }
+            return Ok(user);
+        }
```
Comment (ehimah) [inline src/InstagramClone.Api/Controllers/UsersController.cs:32]: please add comments to each line describing what's happening

File: src/InstagramClone.Api/Startup.cs [hunk 12]
```diff
@@ -13,26 +15,47 @@ public static WebApplication InitialIzeApp(string[] args)
...
         {
+            app.UseHttpsRedirection();
+
+            app.UseRouting();
+
+            
+            app.UseEndpoints(endpoints =>
+            {
+                endpoints.MapControllers();
+            });
```
Comment (ehimah) [inline src/InstagramClone.Api/Startup.cs:51]: please add a line of comment for each line explaining what each does. A better thing will be a line to explain and an external url to the documentation for each

File: webConfig.cs [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+        {
+            // for Attribute roiting
+            config.SuppressDefaultAuthentication();
+            config.MapHttpAttributeRoutes();
+
+            //  conventon based routing
+            config.Routes.MapHttpRoute(
+            name: "DefaultApi",
+            routeTemplate: "api/{controller}/{action}/{id}",
+            defaults: new { id = RouteParameter.Optional }
```
Comment (ehimah) [inline webConfig.cs:20]: same here, we need more docs to add here. each line of `config.` should have an explanation

File: src/InstagramClone.Api/Controllers/UsersController.cs (hunk 11 above)
Comment (sobuse) [reply src/InstagramClone.Api/Controllers/UsersController.cs]: Done

File: src/InstagramClone.Api/Controllers/UsersController.cs [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        public UsersController(InstagramCloneDbContext context)
+        {
+            _context = context;
+        }
+           [HttpGet]
+        [Route("{id:Guid}")]
+        public IHttpActionResult GetUser(Guid id)
+        {
+            // user variable is used to store the value of a particuler user from the database
+            var user = _context.Users.Find(id);
```
Comment (ehimah) [inline src/InstagramClone.Api/Controllers/UsersController.cs:27]: I expect a comment thaty rather explains what is happening don't the right-hand side of the `=`.  Please it should explain `_context.Users.Find(id);`

**Review by ehimah — COMMENTED** [review]
Please make your comments more detailed. It should explain the what and the why.

File: src/InstagramClone.Api/Controllers/UsersController.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sobuse) [reply src/InstagramClone.Api/Controllers/UsersController.cs]: 404 is the response code for not found on the server.

File: src/InstagramClone.Api/Controllers/UsersController.cs (hunk 7 above)
Comment (sobuse) [reply src/InstagramClone.Api/Controllers/UsersController.cs]: Ok, Done

File: src/InstagramClone.Api/Controllers/UsersController.cs (hunk 8 above)
Comment (sobuse) [reply src/InstagramClone.Api/Controllers/UsersController.cs]: I have checked and made the right fix.

File: src/InstagramClone.Api/Startup.cs (hunk 12 above)
Comment (sobuse) [reply src/InstagramClone.Api/Startup.cs]: Ok, Done.

File: src/InstagramClone.Api/Startup.cs [hunk 15]
```diff
@@ -13,26 +15,47 @@ public static WebApplication InitialIzeApp(string[] args)
...
             
+            builder.Services.AddControllers(config =>
+            {
+                config.RespectBrowserAcceptHeader = true;
+                config.ReturnHttpNotAcceptable = true;
+            }).AddXmlDataContractSerializerFormatters();
+
+             builder.Services.AddControllers();
+            builder.Services.AddCors();
+
```
Comment (ehimah) [inline src/InstagramClone.Api/Startup.cs:31]: You didn't add comments here to explain this line

File: src/InstagramClone.Api/Startup.cs (hunk 10 above)
Comment (ehimah) [reply src/InstagramClone.Api/Startup.cs:27]: how about this?

File: src/InstagramClone.Api/Startup.cs (hunk 9 above)
Comment (ehimah) [reply src/InstagramClone.Api/Startup.cs:29]: this is the 2nd time you're calling `builder.Services.AddControllers();` 

the 1st time above has some configuration parameters, and here you call it without any parameters. can you explain the difference?

why do you have it twice?

File: src/InstagramClone.Api/Controllers/UsersController.cs [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+            if (user == null)
+            {   
+                // this give a 404 error responce when the user they searched for by id could not be found 
+                return NotFound();
```
Comment (ehimah) [inline src/InstagramClone.Api/Controllers/UsersController.cs:31]: "response"

File: src/InstagramClone.Api/Controllers/UsersController.cs (hunk 14 above)
Comment (ehimah) [reply src/InstagramClone.Api/Controllers/UsersController.cs:27]: ???

File: src/InstagramClone.Api/Controllers/UsersController.cs (hunk 16 above)
Comment (sobuse) [reply src/InstagramClone.Api/Controllers/UsersController.cs:31]: Ok, Thanks.

File: src/InstagramClone.Api/Controllers/UsersController.cs (hunk 14 above)
Comment (sobuse) [reply src/InstagramClone.Api/Controllers/UsersController.cs]: Ok.Done.

File: src/InstagramClone.Api/Startup.cs (hunk 15 above)
Comment (sobuse) [reply src/InstagramClone.Api/Startup.cs:34]: Ok. Done

**Review by ehimah — APPROVED** [review]
This is awesome

**Pull request merged**